skills/hears.js

```diff
@@ -24,23 +23,38 @@ const extractSkills = (messageString) => {
 };
 
 
-const handler = (bot, message) => {
-
+const handler = async (bot, message) => {
   const messageContent = message.text.replace(LEARNING_KEY, ' ');
-  var skills = extractSkills(messageContent);
-
-  skills.forEach(skill => {
-    Skill.query()
-      .insert({ name: skill })
-      .then(res => bot.reply(message, `${skill} was added as a new skill!`),
-        err => {
-          if (!(err instanceof UniqueViolationError)) {
-            bot.reply(message, `Unable to add ${skill} as a skill :(`);
-          }
-          bot.reply(message, err);
+  const skills = extractSkills(messageContent);
+
+  if (skills.length) {
+    try {
+      const messageRecord = await Message.query().insertAndFetch({
+        text: message.event.text,
+        datetime: new Date().toISOString(),
+      });
+
+      const learnerRecord = await personService.findOrInsertPerson(bot, message.user);
+
+      skills.forEach(async skill => {
+        let skillRecord = await Skill.query().findOne({ name: skill });
+        if (!skillRecord) {
+          skillRecord = await Skill.query().insertAndFetch({ name: skill });
+          bot.reply(message, `${skill} was added as a new skill!`);
         }
-      );
-  });
+
+        // Insert learning point.
+        await Point.query().insert({
+          message_id: messageRecord.id,
+          skill_id: skillRecord.id,
+          teach: false,
+          person_id: learnerRecord.id,
+        });
+      });
+    } catch (err) {
+      console.error(err);
```

Comment:
If we attempt to insert a duplicate skill, I think the `UniqueViolationError` will get caught here, and the learning point will not be awarded. A `UniqueViolationError` is okay - we still want to continue with adding the points. Also, we will have recorded the `Message` record, which will now not be associated with any points. Perhaps consider doing this as a transaction instead

Reply:
> If we attempt to insert a duplicate skill, I think the `UniqueViolationError` will get caught here, and the learning point will not be awarded. A `UniqueViolationError` is okay - we still want to continue with adding the points. Also, we will have recorded the `Message` record, which will now not be associated with any points. Perhaps consider doing this as a transaction instead

@phoebe-lew for each `skill`, it will try to find the skill by name (line 40), and will create a new skill if can't find a record by the name, so I think this will avoid `UniqueViolationError`?
But I agree with `Message` record should be created after get/insert skill success, and if something wrong with find/insert skill neither `Message` nor `Point` record should be created. Let me know if my understanding is wrong.

Reply:
Ah yep, that's correct. Sorry, misread that line!

Yeah that's mostly what I was referring to. In addition to the order of which record gets created in which order, we generally want to avoid undo ALL steps that have happened any fail. Transactions just provide you with an easy way of guaranteeing that without having to manually do "if success then next, if success then next" etc. You can just wrap the whole process (create `Message`, create `Point`, add `Skill`) in a transaction and the whole thing will either succeed or fail.